### PR TITLE
Templatize Point2/PointLL

### PR DESCRIPTION
**Pull request by danpat**
# Issue

The `Point2` class, (and `PointLL` that is subclassed from it) is hard-coded to use `float`.

Most of the time, this is fine.  But sometimes, you want to do something that requires a bit more precision.  In my case, I'm splitting geometry using `midgard::trim_polyline`.

This leads to creating new points along existing lines.  When using `float`, practical precision is limited to about 0.5m wrt geospatial coordinates.  *Most* of the time this is fine.

However, in my case, I'm rendering a map with the original geometry, plus some newly injected intermediate points created by `midgard::trim_polyline`.

When `float` precision is used, sometimes the error margin becomes visible in weird ways, like in the circled section in the following screenshot:

Screen Shot 2020-06-10 at 7 58 31 am

Using `double` for the newly created types has enough precision that they are "on the line" for all human-practical detail levels - `float` does not, and half-metre errors do become visible in some real-world contexts.

Long story short - this PR turns `Point2` into a template class, and makes all the necessary changes to support that.  *Most* code remains unchanged, and `float` precision remains the default, but this refactor allows you to optionally choose `PointXY<double>` or `GeoPoint<double>` (for spherical `Distance`, etc) if you need the extra detail for some reason.

## Tasklist

 - [ ] Add tests
 - [ ] Add #fixes with the issue number that this PR addresses
 - [ ] Generally use squash merge to rebase and clean comments before merging
 - [ ] Update the [changelog](CHANGELOG.md)

## Requirements / Relations

 Link any requirements here. Other pull requests this PR is based on?

### CODE REVIEW

File: valhalla/midgard/vector2.h [hunk 1]
```diff
@@ -14,51 +14,52 @@ namespace midgard {
  * 2D vector class. float x,y components.
  * @author  David W. Nesbitt
  */
-class Vector2 {
+template <typename PrecisionT> class VectorXY {
```
Comment (kevinkreiser) [inline valhalla/midgard/vector2.h:17]: instead of renaming this class and then doing `using Vector2 = VectorXY<float>;` below could we change the default template value to be `float`? ie: `template <typename PrecisionT=float> class Vector2 {`

ah crap. just checked this. seems like pre c++17 you still have to at least put the template operator `Vector2<> foo;` for it to compile. in c++17 and above it will automatically deduce the template argument using the default and doing just `Vector2 foo;` will do the right thing.

Comment (danpat) [reply valhalla/midgard/vector2.h:17]: Unfortunately, yep :-(  I've taken a less invasive approach on a new local branch, still trying to untangle this bit, but it has a lot fewer changes.

Comment (kevinkreiser) [reply valhalla/midgard/vector2.h:17]: roger that, let me know when i should take a look again!

**Comment by danpat** [comment]
@kevinkreiser @PureW This is ready for a round of 👀

File: src/midgard/polyline2.cc [hunk 2]
```diff
@@ -180,6 +16,7 @@ template void
 Polyline2<PointLL>::Generalize(std::list<PointLL>&, float, const std::unordered_set<size_t>&);
 template void
 Polyline2<Point2>::Generalize(std::list<Point2>&, float, const std::unordered_set<size_t>&);
+*/
```
Comment (PureW) [inline src/midgard/polyline2.cc:19]: With these lines commented, there's nothing left in the file. Why not delete it?

Comment (danpat) [reply src/midgard/polyline2.cc]: Ah, forgot to do this one - fixed.

File: valhalla/midgard/distanceapproximator.h [hunk 3]
```diff
@@ -5,7 +5,7 @@
 #include <math.h>
 
 #include <valhalla/midgard/constants.h>
-#include <valhalla/midgard/pointll.h>
+//#include <valhalla/midgard/pointll.h>
```
Comment (PureW) [inline valhalla/midgard/distanceapproximator.h:8]: Nit: cleanup

File: valhalla/midgard/util_core.h [hunk 4]
```diff
@@ -0,0 +1,32 @@
+#pragma once
```
Comment (kevinkreiser) [inline valhalla/midgard/util_core.h:1]: was this pulled out to aid in compilation times or something? i cant quite connect the dots on why its not in `util.h`

File: valhalla/midgard/distanceapproximator.h [hunk 5]
```diff
@@ -28,7 +28,7 @@ namespace midgard {
  * varies with latitude. The distance from A to B may not match the distance
  * from B to A if the latitudes of the 2 points differ.
  */
-class DistanceApproximator {
+template <typename PointT> class DistanceApproximator {
```
Comment (kevinkreiser) [inline valhalla/midgard/distanceapproximator.h:31]: the private members below should have type `typename PointT::PrecisionT`

File: src/midgard/CMakeLists.txt [hunk 6]
```diff
@@ -1,14 +1,10 @@
 file(GLOB headers ${VALHALLA_SOURCE_DIR}/valhalla/midgard/*.h)
 
 set(sources
-  linesegment2.cc
```
Comment (kevinkreiser) [inline src/midgard/CMakeLists.txt:4]: could we keep this out of headers and make explicit template instantiations or does that have another negative side effect that im not thinking of? i cant imagine there is any  other template agrument beyond double that we would care about

File: test/polyline2.cc [hunk 7]
```diff
@@ -52,72 +61,101 @@ TEST(Polyline2, TestGeneralizeSimplification) {
   EXPECT_EQ(line, (Polyline2<Point2>{{{17, 0}, {17, 4}, {17, 5}}}))
       << "Should have removed all but the first, last and marked points";
 
-  Polyline2<PointLL> line1{
```
Comment (kevinkreiser) [inline test/polyline2.cc:55]: its not easy to tell what is different in these tests. is it formatting or did we need to change some of the numbers to get it to work or are there new tests? maybe all 3? just a mini call out as to what is going on here for the lazy among us :D
